kernel/kernel-api/src/main/java/io/delta/kernel/internal/metadatadomain/JsonMetadataDomain.java

```diff
@@ -0,0 +1,113 @@
+/*
+ * Copyright (2024) The Delta Lake Project Authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.delta.kernel.internal.metadatadomain;
+
+import com.fasterxml.jackson.annotation.JsonIgnore;
+import com.fasterxml.jackson.core.JsonProcessingException;
+import com.fasterxml.jackson.databind.*;
+import io.delta.kernel.exceptions.KernelException;
+import io.delta.kernel.internal.SnapshotImpl;
+import io.delta.kernel.internal.actions.DomainMetadata;
+import io.delta.kernel.internal.rowtracking.RowTrackingMetadataDomain;
+import java.util.Optional;
+
+/**
+ * Abstract class representing a JSON metadata domain, whose configuration string is a JSON
+ * serialization of a domain object. This class provides methods to serialize and deserialize a
+ * metadata domain to and from JSON. Concrete implementations, such as {@link
+ * RowTrackingMetadataDomain}, should extend this class to define a specific metadata domain.
+ */
+public abstract class JsonMetadataDomain {
+  // We explicitly set the ObjectMapper to fail on missing properties and unknown properties
+  private static final ObjectMapper OBJECT_MAPPER =
+      new ObjectMapper()
+          .configure(DeserializationFeature.FAIL_ON_MISSING_CREATOR_PROPERTIES, true)
```

Comment:
In Delta Spark, the configuration is 

> _mapper.setSerializationInclusion(Include.NON_ABSENT)
> _mapper.configure(DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES, false)

So it doesn't fail on unknown properties, nor missing properties (default behavior). Serialization also excludes absent values (e.g., empty `Option`/`Optional`). These work together to ensure empty `Option`s are excluded in serialization and are initialized to empty when missing in deserialization.

Shall we do the same thing in Delta Kernel for consistency? 

One concern is whether we should allow missing properties (e.g., initializing `rowIdHighWaterMark` to 0 when it’s absent). This might not be entirely correct. But, with correct implementation, missing properties (except for `Option`) shouldn’t happen. So this may not be an issue.

Curious what do you think, thanks!

Reply:
As discussed, I've configure the `ObjectMapper` with the same settings used in Delta-Spark for consistency and safety.